NETOBSERV-1075 Dashboard feedback: split infra / applications

jpinsonneau opened the pull request:
- moved to embedded json approach instead of FLP for more flexibility
- splitted application / infrastructures namespace & workload graphs
- grouped by sections

![image](https://github.com/netobserv/network-observability-operator/assets/91894519/598b8add-e01f-45bf-9b6e-c4c7ab22656d)
![image](https://github.com/netobserv/network-observability-operator/assets/91894519/fd4d9d32-de76-4a54-a19c-953c88a4d0ee)
![image](https://github.com/netobserv/network-observability-operator/assets/91894519/028bd4c8-5ef1-4479-9730-aa457724e936)

jotak commented on controllers/flowcollector_controller.go:
```diff
@@ -309,6 +309,12 @@ func (r *FlowCollectorReconciler) reconcileOperator(ctx context.Context, cmn *re
 		}
 	}
+
+	desiredFlowDashboardCM := buildFlowMetricsDashboard()
+	if err := cmn.ReconcileConfigMap(ctx, desiredFlowDashboardCM); err != nil {
+		return err
+	}
```
I think it makes sense to move that under the same condition as the health dashboard, in `if r.availableAPIs.HasSvcMonitor()` below? Is there a reason to not do that?

jotak commented on controllers/flowcollector_controller_test.go:
```diff
@@ -326,14 +323,6 @@ func flowCollectorControllerSpecs() {
 					To(ContainElement(v1.Toleration{Operator: v1.TolerationOpExists}))
 			})
-
-			By("Expecting the monitoring dashboards configmap to be deleted")
```
I didn't realize that, but I guess that's the drawback of not using FLP: we loose the ability to tie dashboards with metrics presence. Hence this test not passing, I guess. Can we think about a simple alternative for that? Maybe it requires some discussion?

jpinsonneau replied:
done in https://github.com/netobserv/network-observability-operator/pull/388/commits/71da325c421290cf515e8ef6aed34b6b60db0577 with other enhancements described below

jpinsonneau commented:
Updated this PRs to address all feedback gathered:
- tie metrics to dashboards using tags. Check [FilterDashboardRows](https://github.com/netobserv/network-observability-operator/pull/388/files#diff-72c617dc46a4e4b487ccbed0fd1a1c5a21ec6b9bf0c7cda484076fb35fe13d5e) helper
- added missing metrics definitions
- added / updated missing graphs in dashboards
  -  moved namespace flows to health dashboard
- added a new `health` flag (default) to ignore extra flows graphs per node / namespace / workload

Here are examples of dashboards with no filters:

![image](https://github.com/netobserv/network-observability-operator/assets/91894519/b5447c5c-c605-412d-b74d-d181b428a794)
![image](https://github.com/netobserv/network-observability-operator/assets/91894519/e10c9555-7f00-4868-8bf2-4b8348e03e1b)

jotak commented on api/v1beta1/flowcollector_types.go:
```diff
@@ -318,8 +318,8 @@ type FLPMetrics struct {
-	// Available tags are: `egress`, `ingress`, `flows`, `bytes`, `packets`, `namespaces`, `nodes`, `workloads`.
-	//+kubebuilder:default:={"egress","packets"}
+	// Available tags are: `egress`, `ingress`, `flows`, `bytes`, `packets`, `namespaces`, `nodes`, `workloads`, `health`.
+	//+kubebuilder:default:={"egress","packets", "health"}
```
this would ignore health by default, this isn't what we want?

jpinsonneau replied:
Maybe a better naming can be found ?

The metrics flagged with `health` are only flows `metrics definitions` per nodes / namespaces / workloads

The health dashboard will look like as before by default

jotak replied:
(or health is not what I thought it is )

jotak replied:
It seems like, at least for now, "health" fully overlaps with "flows" so we could keep just "flows"

jotak commented on pkg/helper/dashboard.go:
```diff
@@ -0,0 +1,47 @@
+
+import "encoding/json"
+
+func FilterDashboardRows(dashboard string, ignoreFlags []string) (string, error) {
```
So, if I'm correct, tags are duplicated: in every dashboard row (for dashboards filtering) and in metrics definitions (for FLP confgen). I guess it's ok for now but I think it could be error prone, especially as the dashboard jsons are very noisy and hard to review. So maybe we can think about how would could "secure" this, either for right now or for a follow-up. I can imagine:
- modifying FLP confgen so that it gives us the list of excluded and included metrics; then based on that, we would do the dashboards filtering based on a grep (Contains) on the promql expression
- or do the same thing without confgen, but it means the operator parsing the metrics definition files
- or, since those dashboards are statically defined, we could also just write a unit test that validates that the dashboard tags correctly match the metric defs tags

wdyt?

jotak replied:
btw speaking of unit test I think it would be good to have one here that ensures the dashboard files are correctly parsed without errors

jotak commented on controllers/infra_health_dashboard.json:
```diff
@@ -1,641 +1,1083 @@
+            ],
+            "title": "Flows",
+            "titleSize": "h6"
+        },
+        {
```
If you don't mind me hijacking your PR :-)
I would very well see this kind of addition:
(this is a query that shows the "flows overhead" generated by netobserv itself compared with the total cluster flows. In other words, a way to see if netobserv is triggering a "runaway reaction" as we've already seen in the past, e.g. when we set too small batch sizes.


```suggestion
        },
        {
            "collapse": false,
            "height": "250px",
            "panels": [
                {
                    "aliasColors": {},
                    "bars": false,
                    "dashLength": 10,
                    "dashes": false,
                    "datasource": "prometheus",
                    "fill": 1,
                    "fillGradient": 0,
                    "gridPos": {
                        "h": 20,
                        "w": 25,
                        "x": 0,
                        "y": 0
                    },
                    "id": 3,
                    "legend": {
                        "alignAsTable": false,
                        "avg": false,
                        "current": false,
                        "max": false,
                        "min": false,
                        "rightSide": false,
                        "show": true,
                        "sideWidth": null,
                        "total": false,
                        "values": false
                    },
                    "lines": true,
                    "linewidth": 1,
                    "links": [],
                    "nullPointMode": "null",
                    "percentage": false,
                    "pointradius": 5,
                    "points": false,
                    "renderer": "flot",
                    "repeat": null,
                    "seriesOverrides": [],
                    "spaceLength": 10,
                    "stack": false,
                    "steppedLine": false,
                    "targets": [
                        {
                            "expr": "100 * sum(rate(netobserv_namespace_flows_total{SrcK8S_Namespace='netobserv'}[1m]) or rate(netobserv_namespace_flows_total{SrcK8S_Namespace!='netobserv',DstK8S_Namespace=~'netobserv'}[1m])) / sum(rate(netobserv_namespace_flows_total[1m]))",
                            "format": "time_series",
                            "intervalFactor": 2,
                            "legendFormat": "% overhead",
                            "refId": "A"
                        }
                    ],
                    "thresholds": [],
                    "timeFrom": null,
                    "timeShift": null,
                    "title": "Percentage of flows generated by NetObserv own traffic",
                    "tooltip": {
                        "shared": true,
                        "sort": 0,
                        "value_type": "individual"
                    },
                    "type": "graph",
                    "xaxis": {
                        "buckets": null,
                        "mode": "time",
                        "name": null,
                        "show": true,
                        "values": []
                    },
                    "yaxes": [
                        {
                            "format": "short",
                            "label": null,
                            "logBase": 1,
                            "max": null,
                            "min": null,
                            "show": true
                        },
                        {
                            "format": "short",
                            "label": null,
                            "logBase": 1,
                            "max": null,
                            "min": null,
                            "show": true
                        }
                    ]
                }
            ],
            "repeat": null,
            "repeatIteration": null,
            "repeatRowId": null,
            "showTitle": true,
            "tags": [
                "flows",
                "overhead",
                "flows-overhead"
            ],
            "title": "Flows Overhead",
            "titleSize": "h6"
        },
        {
```

Of course I can also open a separate PR

jpinsonneau replied:
Let's have it ! Thanks

jpinsonneau commented on controllers/flowlogspipeline/metrics_definitions/namespace_flows_total.yaml:
```diff
@@ -8,6 +8,7 @@ usage:
 tags:
   - flows
   - namespaces
+  - health
```
https://github.com/netobserv/network-observability-operator/pull/388/commits/86282d7ef54d0518ef1c50b571fee98208d67585

jotak commented:
@jpinsonneau I pushed some commits:
- to fix bundle/linter issues
- to reduce metrics cardinality related to the new namespace-based metrics; FTR, initially we didn't have specific namespace-based metrics because they are a subset of the existing workload-based metrics. I think adding them makes sense, so that users can disable the "workload" tag and still be able to get namespace-based metrics; but in the default case I think we should not enable both of them (namespaces+workloads) as it is unoptimal => so, my last commit disables the "namespaces" tag and still allow to show namespace-based metrics in the dashboards via the workload metrics.

If any problem with that, let's discuss when you're back

jotak commented:
/lgtm

memodi commented:
/ok-to-test

memodi commented:
@jpinsonneau @jotak - I am not sure what's the intention of "Application" panel here, based on looking at panels, I find difficult to spot differences between Application and Infrastructure.

<img width="1212" alt="image" src="https://github.com/netobserv/network-observability-operator/assets/87338854/fdae4aa8-fd73-4985-91c0-738b5d3a365e">

I was thinking Application would be something that's NOT `netobserv|openshift.*` and to have view something like this:

![image](https://github.com/netobserv/network-observability-operator/assets/87338854/674e67de-3b1a-4b63-a1ad-37baf28d01c6)

wdyt?

jotak commented:
Yes, we should exclude netobserv from applications, I agree

jotak commented:
(FYI I just rebased the branch)

jpinsonneau commented:
Tested back with @jotak changes and works fine:
![image](https://github.com/netobserv/network-observability-operator/assets/91894519/d7bfcaf7-9a72-4c7b-b5ca-a3418bea2780)
![image](https://github.com/netobserv/network-observability-operator/assets/91894519/023fcad7-53f1-480b-a918-ebceed216685)

@memodi note that @jotak excluded netobserv from `Applications` panels
![image](https://github.com/netobserv/network-observability-operator/assets/91894519/ae4de5f4-de40-4b4a-9450-9cd19538822a)
I only see it appearing for kube api queries (default -> netobserv)

jotak commented:
@memodi also another change that I did is to run a top10 instead of top5, because I found top5 has a tendency to often show different metrics for different timestamps, which results in "chunking" more the timeseries. With top10 I feel like the visualization is better / more consistent; sounds good?
